deepspeed/runtime/zero/stage3.py

```diff
@@ -1500,7 +1500,8 @@ def partition_grads(self, params_to_release: List[Parameter], grad_partitions: L
                     else:
                         fp32_grad_tensor = self.fp32_partitioned_groups_flat[i].grad.narrow(
                             0, dest_offset, grad_buffer.numel())
-                        fp32_grad_tensor.copy_(grad_buffer)
+                        fp32_grad_tensor.copy_(
+                            grad_buffer.to(dtype=torch.float32, device=self.device, non_blocking=True).pin_memory())
```

Comment:
Yes, I agree that on-the-fly pinned memory may cause issues and thank you for pointing it out. I've updated the code accordingly.

Reply:
Thanks for making the change. I think there should be a corresponding H2D since grad accum computation is done on device. Can please you check for this?

Reply:
The corresponding H2D is on [L1479-1481 of stage3](https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/zero/stage3.py#L1479-L1481). 
It will 
1) H2D copy, 
2) data type cast on GPU ,
3) do the addition ,
4) D2H copy, 

so the data type cast will only happens on the GPU.

Reply:
@xylian86, thanks for getting this answer. Can you investigate whether doing H2D copy in 16bit as opposed to 32bit will be faster because of lower transfer volume. This will require data type cast on CPU into pinned buffer that you just added. Thanks!

Reply:
Got it. If I understand correctly,  let's assume source is FP32 tensor on CPU, the target is FP16 tensor on GPU, we have two ways to do the H2D copy: 
Option 1) cast to FP16 on CPU(pin) memory, and copy to GPU 
Option 2) copy FP32 tensor to GPU, cast to FP16 on GPU

---
**Results (average for 100 iterations, on H200):**
_Size: 1MB_
Option1: cast on CPU: 1.24ms, copy FP16 to GPU 0.05ms, total: 1.29ms
Option2: copy FP32 to GPU 0.02ms, cast on GPU 0.04ms, total: 0.06ms

_Size: 10MB_
Option1: cast on CPU: 1.28ms, copy FP16 to GPU 0.08ms, total: 1.36ms
Option2: copy FP32 to GPU 0.05ms, cast on GPU 0.04ms, total: 0.09ms

_Size: 100MB_
Option1: cast on CPU: 2.19ms, copy FP16 to GPU 0.32ms, total: 2.51ms
Option2: copy FP32 to GPU 0.29ms, cast on GPU 0.18ms, total: 0.47ms

_Size: 1024MB_
Option1: cast on CPU: 92.16ms, copy FP16 to GPU 1.66ms, total: 93.83ms
Option2: copy FP32 to GPU 2.80ms, cast on GPU 1.26ms, total: 4.05ms

---
The option2 is always faster across different sizes, but the results also depend on the bandwidth between CPU and GPU, if the bandwidth is limited, then Option1 may be better in some cases. Also the Option2 requires more GPU memory

Reply:
The cast-on-CPU is slower than expected, considering that this pure memory transfer and should be bandwidth-bound. The fastest case is 100MB, with achieved throughput of ~50GB/sec. which I suspect is much lower than DRAM bandwidth. Do you know the DRAM b/w of the H200?

Reply:
The DRAM b/w of H200 in theoretical is 500GB/s. I tested more rounds, for 100MB, the fastest cast is 1.6ms.

Reply:
Got it. Let's make note to investigate faster CPU-casts in future. Current solution is fine. Thanks.